### ISSUE TITLE
Pretty-printing for string interpretations of BitArray

### ISSUE THREAD

**Issue by obarak** [scott-griffiths/bitstring#231]
### I'd like to propose adding pretty-printing for the string representations (`.bin`, `.hex`, `.oct`). ###
Pretty-printing should include:
* **Character grouping (separator between every *n* characters)**:
  * Python's own string formatter groups every *n=4* characters using `'_'` as separator (starting from right, i.e. decimal 65 is `'100_0001'` and not `'1000_001'`). So that seems like a good (and predictable) default to use.
  * An advantage of '_' is that Python 3.6+ accepts it in integer literals, so `0b1000_001` is a valid number.
  * Other common options for separator can be whitespace or dot. I think that for simplicity just choose one and don't leave it as a parameter.
  * Similarly, *n=4* should work well across representations, but some may argue for *n=8* in `bin` and *n=2* in `hex`. So...
  * If it's desirable for the parameters to be controlled, Python's [standard library](https://docs.python.org/3/library/stdtypes.html#bytes.hex) uses `bytes.hex(sep, bytes_per_sep)`, where a negative `bytes_per_sep` (in our case `chars_per_sep`) implies starting grouping from the left.

* **Line wrapping and indentation**:
  * More of a nice-to-have, but this should wrap the lines (at some multiplier of `n`) and indent the next line accordingly (so the right-most digits of each line will be aligned).
  * This is similar to what's done in Python's [pretty-print](https://docs.python.org/3/library/pprint.html).

**Comment by scott-griffiths** [comment]
It's a nice idea. In practice it's replacing `a.bin` with `f'{a.uint:_b}'` I think:

```
>>> a = Bits(uint=10000, length=16)
>>> a.bin
'0010011100010000'
>>> f'{a.uint:_b}'
'10_0111_0001_0000'
```

Except that loses the leading zeros...
Similarly with `_o` and `_x` for octal and hexadecimal.

Amusingly this doesn't work:

```
>>> f'{a.uint:0>len(a)_b}'
'10_0111_0001_0000'
```

As the length of 16 includes the added underscores! Which leaves the horrible `f'{a.uint:0>{len(a) + (len(a) - 1) // 4}_b}'`. 

I'll have to take a look at what pretty-print does. Whatever's printed this needs to still work:

```
>>> a.bin
'0_1001 ...
...
... _1011'
>>> b = Bits(bin=_)
>>> assert a == b
```

**Comment by obarak** [comment]
Great!



Re:
 > Whatever's printed this needs to still work

You'll probably need to add support for underscores in the string initializers as well.

**Comment by scott-griffiths** [comment]
Having thought about it a bit more I can't implement this as described. The problem is that a lot of code depends on the string representation as the input for further calculation so messing with it will break a lot of user code (and indeed my testing code!)

One possibility would be something like `bitstring.pprint(a.bin)` but the function in this case wouldn't know the type of thing it was printing (it just gets a string, and doesn't know it's binary).

Another possibility is `a.print('bin')` with the `print` method taking similar options to the pretty print module's options. This could then be a pure printing option so could include line breaks and other formatting.

```
def print(self, format='bin', width=80, group_size=8, separator=' ', show_offset=False, stream=None, zero_pad=True):
"""
    Print the bitstring to stream (defaults to sys.stdout) with human readable formatting.

    format - One of 'bin', 'oct', 'hex'. (Note: I don't think we can do 'bytes')
    width - the desired maximum number of characters per line in the output. Can be set to 0 or None to have no line breaks.
    group_size - the number of characters in each printed group
    separator - a string to use between printed groups. Can be set to '' or None to have no separation
    show_offset - show bit offset at start of each line. Only useful if width is also set.
    stream - (default sys.stdout) is a file-like object to which the output will be written by calling its write() method.
    zero_pad - Option to pad with zeros for oct and hex to make them up to a whole number of characters. Is this padding at start or end? Not sure I like it on second thoughts - maybe just throw an error!
"""
```

```
>>> a = Bits(1000)
>>> print(a.bin)
0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
>>> a.print('bin', show_offset=True)
   0 : 00000000 00000000 00000000 00000000 00000000 00000000
  48 : 00000000 00000000 00000000 00000000 00000000 00000000
  96 : 00000000 00000000 00000000 00000000 00000000 00000000
 144 : 00000000 00000000 00000000 00000000 00000000 00000000
...
```

**Comment by obarak** [comment]
An object `print` method sounds good and the `show_offset` option is *brilliant*!

Maybe shorten `print` to `pp`, like Python's pprint does, or even make the default setting a property `.pbin/.phex/.poct` (since I see myself using it all the time in an interactive session, and having additional parens to type actually makes a difference 😅). 

`format` - maybe you can special-case `bytes` to print hex with a space between every two characters (on top of the other grouping options), making it similar to a [memory viewer view](https://www.nirsoft.net/utils/heapmemview.gif).

`zero-pad` - how about `left` (default) and `right` instead of boolean? It would be annoying to have to manually compute padding every time just for printing.

Now I'm knit-picking, but I would change `separator` to `sep`, like in other std. library functions.

**Comment by scott-griffiths** [comment]
Thanks for the input. I was considering whether to call it `pprint` but maybe `pp` is better. Agree that `sep` is more standard (and shorter).

For the hex format you can use

```
a.pp('hex', group_size=2)
```

to get the effect you want. I'm thinking of expanding the `format` option slightly to allow more than one view. So you could do

```
>>> a.pp('bin, hex')
   0 : 00000000 00000000 00000000 00000000 00000000  00 00 00 00 00
  48 : 00000000 00001111 00000000 00000000 00000000  00 0f 00 00 00
  96 : 00000000 00000000 11110000 00000000           00 00 f0 00
```

That would also work well with `'bytes'` which I think I'll also add (replacing lots of unprintable characters with `.` seems usual).

I'd already abandoned the `zero_pad` option for an `align_left` bool.

**Comment by obarak** [comment]
Thanks, looks good.

What do you think about the `set_lsb0` support? 
You'll still print the left-most bit first (so, internally you need the real `v[0]`, not the `set_lsb0`'s `v[0]`), and then tweak the offset numbering.

**Comment by scott-griffiths** [comment]
The `lsb0` stuff makes my brain hurt. It all sounds so simple, but in practice I haven't managed to finish that piece of work yet despite making a few attempts.

So how would the offset numbering work? For example if we had

```
>>> a.pp()
  0: 00000000 11111111
 16: 00001111 11110000
 32: 1010
```
Then if we `set_lsb0 = True`:

```
>>> a.pp()
 32: 0000
 16: 00001111 11110000
  0: 11111111 00001010 
```
which is sort of OK, but I think we really want to display the lines in the opposite order too:

```
>>> a.pp()
  0: 11111111 00001010 
 16: 00001111 11110000
 32: 0000
```

So I think it's fine, but will probably not make the first version.

**Comment by obarak** [comment]
### It might be simpler than that (I think the discussion in #156 was overly confusing). ###

The important thing is that lsb0 only refers to how you number the bits, that's it. 

I'm not familiar with all use-cases, but lsb0 is the de-facto standard in documenting hardware data structures (e.g. CPU registers), and I believe that's where the request came from.

And all documentation I know of (Intel, Amd, ARM...) print the MSB as usual - in the left - **with lsb0 again referring only to the bit numbering**. 

e.g. the diagram below (from an Intel manual):
![image](https://user-images.githubusercontent.com/6918300/181618249-660375e7-1664-491f-b967-90f7f0162a2b.png)

The manual would say "Family ID is bits[11:8]". And they use [11:8] and not [8:11] to mean "take bits 8 thru 11 (in lsb0 numbering), but still treat bit 11 as the MSB".

To sum up, the way I see it, `lsb0_mode` is only for convenience, so I could do `a[8:11]` like the manual says and get the right field.

Re your example, I don't think we need to show the lines in opposite order. The MSB is still first, only its index is different. What I would do differently is change the offset labels. They should either be on the right side of the line (because that's where bit number 0 starts) or be on the left but show the number of the left bit in the line.

``` python
# default mode
>>> a.pp()
  0: 00000000 11111111
 16: 00001111 11110000
 32: 1010

# lsb0 mode option 1
>>> a.pp()
 0000               :32
 00001111 11110000  :16
 11111111 00001010  :0

# lsb0 mode option 2
>>> a.pp()
 35: 0000
 31: 00001111 11110000
 15: 11111111 00001010 
```

**Comment by scott-griffiths** [comment]
I think this is now finished on the 4.0-update branch.
